src/lib.rs

```diff
@@ -105,47 +115,34 @@ impl<Data: AsRef<[u8]>> Reader<Data> {
         &self.input.as_ref()[start..end]
     }
 
-    pub fn data_format_descriptors(&self) -> ParseResult<impl Iterator<Item = DataFormatDescriptor>> {
+    pub fn data_format_descriptors(&self) -> impl Iterator<Item = DataFormatDescriptor> {
         let header = self.header();
         let start = header.dfd_byte_offset as usize;
-        let length: u32 = u32::from_le_bytes(
-            self.input
-                .as_ref()
-                .get(start..start + 4)
-                .ok_or(ParseError::UnexpectedEnd)?
-                .try_into()
-                .unwrap(),
-        );
-        if length != header.dfd_byte_length || length != header.kvd_byte_offset - header.dfd_byte_offset {
-            return Err(ParseError::UnexpectedEnd);
-        }
         let end = (header.dfd_byte_offset + header.dfd_byte_length) as usize;
-        Ok(DataFormatDescriptorIterator {
-            data: self.input.as_ref(),
-            offset: start + 4,
-            end,
-        })
+        DataFormatDescriptorIterator {
+            data: &self.input.as_ref()[start..end],
+            offset: 0,
+        }
     }
 }
 
 struct DataFormatDescriptorIterator<'data> {
     data: &'data [u8],
     offset: usize,
-    end: usize,
 }
 
 impl<'data> Iterator for DataFormatDescriptorIterator<'data> {
     type Item = DataFormatDescriptor<'data>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        if self.offset > self.end - DataFormatDescriptorHeader::LENGTH {
+        if self.offset > self.data.len() - DataFormatDescriptorHeader::LENGTH {
```

Comment:
Rather than tracking `offset` separately, could we just advance the start of `data`?

Reply:
Nice. Done.